### DISCUSSION TITLE
How do I get the pathname with hash.

### DISCUSSION THREAD

**Discussion by brownjedi** [vercel/next.js#49465] [Help] [answered]
### Summary

Hello! I would like to have access to the URL in server side components. I'm using the `usePathname` from `next/navigation`, but it's stripping the hash. Is there a way to access the hash?

**Comment by icyJoseph** [comment, marked as answer]
The fragment part of a URL, also known as anchor, is not part of the path. If we look at the `window.location` API:

- https://developer.mozilla.org/en-US/docs/Web/API/Location#location_anatomy
- https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Web_mechanics/What_is_a_URL

We see that it is a distinct part of the URL. Much like, search params, protocol, etc. 

There's one more thing to consider with the fragment/anchor/hash. It is not shared with the server. When you make a request, the browser removes it from the url, so a server component won't ever see it, unless you share it as queries, or within the body of a POST request, for instance. 

Also, `usePathname` won't update when the `window.location.hash`, changes. 

### Using the hashchange event

-  https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event

### Hack

I think you can do a little hack though... with `useParams`:

> Remember that these are Next.js Client Components, and must account for window being `undefined`, during SSR. That's why I do `window.location.hash` inside an effect.

From the docs:

> Components in the Client Component module graph are primarily rendered on the client, but with Next.js, they can also be pre-rendered on the server and hydrated on the client.


And now the code:

```tsx
"use client";

import { useRouter, useParams } from "next/navigation";
import { useEffect } from "react";

export default function Home() {
  const router = useRouter();
  const params = useParams();

  useEffect(() => {
    console.log("Hash:", window.location.hash);
  }, [params]);

  const pushRandomUUID = () => router.push(`/#${window.crypto.randomUUID()}`);

  return (
    <main>
      <button onClick={pushRandomUUID}>Push fragment</button>
    </main>
  );
}

```

This one will log to the client the current hash. If there's ever an improvement to the params object, such that it bails out, when no meaningful change happened, then this hook will stop working as intended.

Alternatively you could just read the value of `window.location.hash` when you need it. Though it might lead to impure rendering. Callbacks and event handlers should be fine.

**Reply by DNA-PC** [reply]
Kinda both up/down voted
Thanks for the solution, it's just 🤯 not intuitive

**Reply by pedroSoaresll** [reply]
This strategy works on version `15.0.0-canary.0` but not on version `15.0.1`.
The version `15.0.2-canary.1` is not working too.

**Reply by ghost** [reply]
I had the same problem, was able to get similar results using useSearchParams instead of useParams.

**Comment by inadeqtfuturs** [comment]
This would be really useful to add to the official documentation -- getting the most recent hash by combining hooks.

**Comment by monolithed** [comment]
The workaround above will function correctly only if the user opens the page through a direct link. 
For internal navigations, the router's state will change faster than the actual hash in the address bar. 
It's quite puzzling that there isn't a proper way to retrieve the hash, as it's crucial for anchor navigation within the page.

**Comment by andelkocvjetkovic** [comment]
I encountered the same issue and decided to switch to using a native anchor element for my solution. Here's the code:

```tsx
<a href='#go-to'>Go to</a>
<h1 id='go-to'>Hello world</h1>
```

I then created a custom hook called `useHash`:

```tsx
'use client';
import { useEffect, useState } from 'react';

const getHash = () => (typeof window !== 'undefined' ? decodeURIComponent(window.location.hash.replace('#', '')) : undefined);

const useHash = () => {
  const [hash, setHash] = useState(getHash());

  useEffect(() => {
    const handleHashChange = () => {
      setHash(getHash());
    };
    window.addEventListener('hashchange', handleHashChange);
    return () => {
      window.removeEventListener('hashchange', handleHashChange);
    };
  }, []);

  return hash;
};

export default useHash;
```

You can use it in your code like this:

```tsx
const hash = useHash();
```

I believe this is a good solution because the anchor element is used for scrolling behavior rather than navigating across the app.

**Reply by MarJaysonSanAgustin** [reply]
Thanks for this.

I have just added: 
```typescript
const [isClient, setIsClient] = useState(false);
```
to prevent this nextjs hydration error: `https://nextjs.org/docs/messages/react-hydration-error`

here is the complete code:
```typescript
"use client";
import { useEffect, useState } from "react";

const getHash = () =>
  typeof window !== "undefined"
    ? decodeURIComponent(window.location.hash.replace("#", ""))
    : undefined;

const useHash = () => {
  const [hash, setHash] = useState(getHash());
  const [isClient, setIsClient] = useState(false);

  useEffect(() => {
    setIsClient(true);
    const handleHashChange = () => {
      setHash(getHash());
    };
    window.addEventListener("hashchange", handleHashChange);
    return () => {
      window.removeEventListener("hashchange", handleHashChange);
    };
  }, []);

  return isClient ? hash : null;
};

export default useHash;
```

**Reply by misha-erm** [reply]
not sure why but seems this doesn't work with app router. 'hashchange' never gets called after I do `router.push('#new-hash')`

UPDATE: ookay, seems like the problem was in my "event.preventDefault" inside click handler

**Reply by zoul** [reply]
Thanks! I believe you also need a `setHash(getHash())` before adding the event listener to initialize the hash, otherwise the hook didn’t work for me for SPA navigation.

**Reply by kevludwig** [reply]
The event listener doesn't work in app-dir context. You can listen to changes from useParams and rerender on change. 

Here is a minimum working example:

```jsx
"use client";

import { useParams } from "next/navigation";
import { useEffect, useState } from "react";

const getHash = () =>
  typeof window !== "undefined" ? window.location.hash : undefined;

const useHash = () => {
  const [isClient, setIsClient] = useState(false);
  const [hash, setHash] = useState(getHash());
  const params = useParams();

  useEffect(() => {
    setIsClient(true);
    setHash(getHash());
  }, [params]);

  return isClient ? hash : null;
};

export default useHash;
```

**Reply by ljrahn** [reply]
event listener also didnt work for me, but this worked, thanks!

**Reply by webbertakken** [reply]
Works great. I've adapted it a bit to also remove hashes using the same hook.

Here's a minimum working example:

```ts
'use client'

import { useParams, usePathname, useRouter, useSearchParams } from 'next/navigation'
import { useEffect, useState } from 'react'

const getHash = () => (typeof window !== 'undefined' ? window.location.hash : undefined)

/**
 * based on https://github.com/vercel/next.js/discussions/49465#discussioncomment-7968587
 */
export function useHash() {
  const params = useParams()
  const path = usePathname()
  const search = useSearchParams()
  const router = useRouter()

  const [isClient, setIsClient] = useState(false)
  const [hash, setHash] = useState(getHash())

  useEffect(() => {
    setIsClient(true)
    setHash(getHash())
  }, [params])

  const removeHash = () => {
    setHash('')
    let newPath = path
    if (search) newPath += `?${search.toString()}`
    router.replace(newPath, { scroll: false })
  }

  return { hash: isClient ? hash : null, removeHash }
}
```

**Reply by behzad-ahmadi** [reply]
> The event listener doesn't work in app-dir context. You can listen to changes from useParams and rerender on change.
> 


Thanks. Here is an improved version.
For me returning null raises Not Found Boundary error.
Next.js 14.2

```
'use client'

import { useParams } from 'next/navigation'
import { useEffect, useState } from 'react'

const getHash = () =>
  typeof window !== 'undefined' ? window.location.hash : ''

const useHash = () => {
  const [isClient, setIsClient] = useState(false)
  const [hash, setHash] = useState<string>(getHash())
  const params = useParams()

  useEffect(() => {
    setIsClient(true)
    setHash(getHash())
  }, [params])

  return isClient ? hash : ''
}

export default useHash

```

**Comment by SamDeimos** [comment]
This works for me

```javascript
"use client"
import { useParams } from "next/navigation"
import { useEffect, useState } from "react"

const useHash = () => {
  const params = useParams()
  const [hash, setHash] = useState()

  useEffect(() => {
    const currentHash = window.location.hash.replace("#", "")
    setHash(currentHash)
  }, [params])

  return hash
}

export default useHash
```

**Comment by prince272** [comment]
This also works for me, supports both next routing hash changes and native browser hash changes.

You will need to install: npm i "query-string"
```` jsx
"use client";

import { useEffect, useMemo, useState } from "react";
import { useParams, useRouter } from "next/navigation";
import queryString from "query-string";

// How do I get the pathname with hash.
// source: https://github.com/vercel/next.js/discussions/49465
export const useHashState = () => {
  const getCurrentHash = useMemo(() => () => (typeof window !== "undefined") ? window.location.hash.replace(/^#!?/, "") : "", []);
  const router = useRouter();
  const params = useParams();
  const [hash, _setHash] = useState<string>(getCurrentHash());

  const setHash = (newHash: string) => {
    let updatedUrl = window.location.href;
    updatedUrl = queryString.stringifyUrl({ url: updatedUrl.split('#')[0], fragmentIdentifier: newHash });

    _setHash(newHash);
    router.replace(updatedUrl);
  };
  useEffect(() => {
    const currentHash = getCurrentHash();
    _setHash(currentHash);
  }, [params]);

  const handleHashChange = () => {
    const currentHash = getCurrentHash();
    _setHash(currentHash);
  };

  useEffect(() => {
    window.addEventListener("hashchange", handleHashChange);

    return () => {
      window.removeEventListener("hashchange", handleHashChange);
    };
  }, []);

  return [hash, setHash] as const;
};

````

**Reply by ZambranoTech** [reply]
thx very helpful

**Reply by aniketbhalla2544** [reply]
worked for me too, in **next.js 14**, but I just did some minor changes, like i removed useMemo and took getCurrentHash out from the component, took handleHashChange inside the "listener useEffect".

```
"use client";

import { useEffect, useState } from "react";
import { useParams, useRouter } from "next/navigation";
import queryString from "query-string";

const getCurrentHash = () =>
  typeof window !== "undefined" ? window.location.hash.replace(/^#!?/, "") : "";

export const useHashState = () => {
  const router = useRouter();
  const params = useParams();
  const [hash, _setHash] = useState<string>(getCurrentHash());

  const setHash = (newHash: string) => {
    let updatedUrl = window.location.href;
    updatedUrl = queryString.stringifyUrl({
      url: updatedUrl.split("#")[0],
      fragmentIdentifier: newHash,
    });

    _setHash(newHash);
    router.replace(updatedUrl);
  };
  useEffect(() => {
    const currentHash = getCurrentHash();
    _setHash(currentHash);
  }, [params]);

  useEffect(() => {
    const handleHashChange = () => {
      const currentHash = getCurrentHash();
      _setHash(currentHash);
    };

    window.addEventListener("hashchange", handleHashChange);

    return () => {
      window.removeEventListener("hashchange", handleHashChange);
    };
  }, []);

  return [hash, setHash] as const;
};

```

**Comment by MuhammadHaseeb12** [comment]
For React and Next js you can use very light package that i have build  called r[each-hash-control ](https://www.npmjs.com/package/react-hash-control) related to the url hash to read url hash params and get hash history and to scroll between hashes with in the page and more. check out https://github.com/haseeb5555/react-hash-control (give star if you find helpfull)

**Comment by billy-the-ape** [comment]


**Reply by icyJoseph** [reply]
> **Is it possible at all to read the hash using the app router during SSR??**
> 
> Wild that these workarounds have to exist -- and then we still get the ` Hydration failed because the initial UI does not match what was rendered on the server.` error if we actually USE the hash....

Browsers don't send the hash fragment to the server. Look it up!

For example:

- https://stackoverflow.com/questions/317760/how-to-get-url-hash-from-server-side

**Comment by Andrew-Chen-Wang** [comment]
Thank you to this SO answer for NextJS 15 that grabs the hash AND grabs the hash changes.

https://stackoverflow.com/a/79215367/12875823

```typescript
'use client'

import React from 'react'

export const useHash = () => {
    const [hash, setHash] = React.useState(window.location.hash)
    React.useEffect(() => {
        const onHashChanged = () => setHash(window.location.hash)
        const { pushState, replaceState } = window.history
        window.history.pushState = function (...args) {
            pushState.apply(window.history, args)
            setTimeout(() => setHash(window.location.hash))
        }
        window.history.replaceState = function (...args) {
            replaceState.apply(window.history, args)
            setTimeout(() => setHash(window.location.hash))
        }
        window.addEventListener('hashchange', onHashChanged)
        return () => {
            window.removeEventListener('hashchange', onHashChanged)
        }
    }, [])
    return hash
}
```

**Reply by webbertakken** [reply]
Seems solid. Thanks for sharing!

I've adapted my version (from [above](https://github.com/vercel/next.js/discussions/49465#discussioncomment-10040050)), that includes a `removeHash` method to be based on this.

<details>
<summary>Diff</summary>

```diff
'use client'

import React from 'react'
+ import { usePathname, useRouter, useSearchParams } from 'next/navigation'

export const useHash = () => {
+   const router = useRouter()
+   const path = usePathname()
+   const search = useSearchParams()

...

+   const removeHash = () => {
+     let newPath = path ?? ''
+     if (search) newPath += `?${search.toString()}`
+     router.replace(newPath, { scroll: false })
+   }

-   return hash
+   return { hash, removeHash }
}
```

</details>

<details>
<summary>Full version</summary>

```ts
'use client'

import React from 'react'
import { usePathname, useRouter, useSearchParams } from 'next/navigation'

export const useHash = () => {
  const router = useRouter()
  const path = usePathname()
  const search = useSearchParams()

  const [hash, setHash] = React.useState(window.location.hash)

  React.useEffect(() => {
    const onHashChanged = () => setHash(window.location.hash)
    const { pushState, replaceState } = window.history
    window.history.pushState = function (...args) {
      pushState.apply(window.history, args)
      setTimeout(() => setHash(window.location.hash))
    }
    window.history.replaceState = function (...args) {
      replaceState.apply(window.history, args)
      setTimeout(() => setHash(window.location.hash))
    }
    window.addEventListener('hashchange', onHashChanged)
    return () => {
      window.removeEventListener('hashchange', onHashChanged)
    }
  }, [])

  const removeHash = () => {
    let newPath = path ?? ''
    if (search) newPath += `?${search.toString()}`
    router.replace(newPath, { scroll: false })
  }

  return { hash, removeHash }
}
```

</details>
